torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,134 @@ namespace rpc {
 
 constexpr size_t WorkerInfo::MAX_NAME_LEN;
 
+// TODO(osalpekar): Tune RetryDuration to be something in seconds per
+// team recommendations. Kept at 100ms for the sake of Milestone 1.
+const auto kRpcRetryDuration = std::chrono::milliseconds(100);
+// kBackoff is the multiplier for the exponential backoff algorithm.
+const float kBackoff = 1.2;
+
 RpcAgent::RpcAgent(
     WorkerInfo workerId,
     std::unique_ptr<RequestCallback> cb,
     std::chrono::milliseconds rpcTimeout)
     : workerInfo_(std::move(workerId)),
       cb_(std::move(cb)),
       rpcTimeout_(rpcTimeout),
-      profilingEnabled_(false) {}
+      profilingEnabled_(false) {
+  {
+    std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+    rpcAgentRunning_.store(true);
+  }
+  rpcRetryThread_ = std::thread(&RpcAgent::retryExpiredRpcs, this);
+}
+
+RpcAgent::~RpcAgent() {
+  {
+    std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+    rpcAgentRunning_.store(false);
+  }
+  rpcRetryMapCV_.notify_one();
+  rpcRetryThread_.join();
+}
 
-RpcAgent::~RpcAgent() = default;
+std::shared_ptr<FutureMessage> RpcAgent::sendWithRetries(
+    const WorkerInfo& to,
+    Message&& message,
+    int maxRetries) {
+  std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+  auto future = std::make_shared<FutureMessage>();
+  auto fm = send(to, std::move(message));
+  fm->addCallback([future](
+                      const rpc::Message& message,
+                      const c10::optional<utils::FutureError>& futErr) {
+    if (futErr) {
+      LOG(INFO) << "Send try 1 failed\n";
+    } else {
+      // First try succeeded and we can mark the original future as completed.
+      future->markCompleted(message);
+    }
+  });
+  steady_clock_time_point newTime =
+      std::chrono::time_point_cast<std::chrono::milliseconds>(
+          std::chrono::steady_clock::now() + kRpcRetryDuration);
+  rpcRetryMap_.emplace(
+      std::piecewise_construct,
+      std::forward_as_tuple(newTime),
+      std::forward_as_tuple(RpcRetryInfo(
+          to, std::move(message), fm, future, /* retryCount */ 1, maxRetries)));
+  rpcRetryMapCV_.notify_one();
+
+  return future;
+}
+
+void RpcAgent::retryExpiredRpcs() {
+  while (rpcAgentRunning_.load()) {
+    // TODO(osalpekar): This is temporary solution for handling empty map and
+    // sleeping until RPC retry duration is expired per Milestone 1.
+    std::unique_lock<std::mutex> lock(rpcRetryMutex_);
+    rpcRetryMapCV_.wait(lock, [&, this] {
+      return !(rpcAgentRunning_.load() && rpcRetryMap_.empty());
+    });
+
+    if (!rpcAgentRunning_.load()) {
+      return;
+    }
+    steady_clock_time_point earliestTimeout = rpcRetryMap_.begin()->first;
+    if (std::chrono::steady_clock::now() < earliestTimeout) {
+      std::this_thread::sleep_until(earliestTimeout);
```

Comment:
pls don't sleep here,  since then the destructor will block for an arbitrary time period.

you should instead use the rpcRetryMapCV_ to do the sleeping above (e.g. calculate a timeout, then cv.wait_for()